src/typing/typecheck.ml

```diff
@@ -562,76 +597,84 @@ let add_tydef env (t, (params, def)) =
+    | Dsyntax.ML_Sum constructors ->
        let constructors = List.map (fun (c, ts) -> c, List.map (ml_ty params) ts) constructors in
        Tyenv.topadd_tydef t (Mlty.Sum (params, constructors)) env
 
-let add_operation op (args, out) env =
-  let args = List.map (ml_ty []) args
-  and out = ml_ty [] out in
-  Tyenv.topadd_operation op (args, out) env
-
-let rec toplevel env ({Location.thing=c; loc} : _ Syntax.toplevel) =
+let rec toplevel env ({Location.thing=c; loc} : Dsyntax.toplevel) =
   match c with
   (* Desugar is the only place where recursion/nonrecursion matters *)
-  | Syntax.DefMLType tydefs ->
+  | Dsyntax.DefMLType tydefs ->
     let env = List.fold_left add_tydef env tydefs in
-    env, locate ~loc (Syntax.DefMLType tydefs)
+    env, locate ~loc (Rsyntax.DefMLType tydefs)
 
-  | Syntax.DefMLTypeRec tydefs ->
+  | Dsyntax.DefMLTypeRec tydefs ->
     let env = List.fold_left add_tydef env tydefs in
-    env, locate ~loc (Syntax.DefMLTypeRec tydefs)
+    env, locate ~loc (Rsyntax.DefMLTypeRec tydefs)
 
-  | Syntax.DeclOperation (op, opty) ->
-    let env = add_operation op opty env in
-    env, locate ~loc (Syntax.DeclOperation (op, opty))
+  | Dsyntax.DeclOperation (op, (tys_in, ty_out)) ->
+    let tys_in = List.map (ml_ty []) tys_in in
+    let ty_out = ml_ty [] ty_out in
+    let env = Tyenv.topadd_operation op (tys_in, ty_out) env in
+    env, locate ~loc (Rsyntax.DeclOperation (op, (tys_in, ty_out)))
 
-  | Syntax.DeclConstants (cs, t) ->
-    let env, t = Tyenv.at_toplevel env (check_comp t Mlty.Jdg) in
-    env, locate ~loc (Syntax.DeclConstants (cs, t))
+  | Dsyntax.DeclConstants (cs, t) ->
+    let env, t = Tyenv.at_toplevel env (check_comp t Mlty.Judgment) in
+    env, locate ~loc (Rsyntax.DeclConstants (cs, t))
 
-  | Syntax.TopHandle lst ->
+  | Dsyntax.TopHandle lst ->
     let env, lst = Tyenv.at_toplevel env (top_handler ~loc lst) in
-    env, locate ~loc (Syntax.TopHandle lst)
+    env, locate ~loc (Rsyntax.TopHandle lst)
 
-  | Syntax.TopLet xcs ->
+  | Dsyntax.TopLet xcs ->
     let env, xcs = Tyenv.at_toplevel env (let_clauses xcs) in
     let env = List.fold_left (fun env (x, s, _) -> Tyenv.topadd_let x s env) env xcs in
-    env, locate ~loc (Syntax.TopLet xcs)
+    env, locate ~loc (Rsyntax.TopLet xcs)
 
-  | Syntax.TopLetRec xycs ->
+  | Dsyntax.TopLetRec xycs ->
     let env, xycs = Tyenv.at_toplevel env (let_rec_clauses xycs) in
     let env = List.fold_left (fun env (x, _, s, _) -> Tyenv.topadd_let x s env) env xycs in
-    env, locate ~loc (Syntax.TopLetRec xycs)
-
-  | Syntax.TopDynamic (x, _, c) ->
-    let env, (c, t) = Tyenv.at_toplevel env (comp c) in
-    let s = Mlty.ungeneralized_schema t in
-    let env = Tyenv.topadd_let x s env in
-    env, locate ~loc (Syntax.TopDynamic (x, s, c))
+    env, locate ~loc (Rsyntax.TopLetRec xycs)
+
+  | Dsyntax.TopDynamic (x, annot, c) ->
+    let env, (c, t) =
+      Tyenv.at_toplevel env
+        (comp c >>= fun (c, t) ->
```

Comment:
I sort of feel like this should be its own function, like check_annotated_comp or something, but I'm not sure.
If this comment inspires you please make a followup PR.